README.md
# OpenFaaS AMQP 1.0 Connector

An OpenFaaS connector for working with AMQP 1.0

## AMQP Services Supported

This has been tested with the following services, but should work with any message queue that follows the 
[AMQP v1.0](https://www.amqp.org/resources/specifications) standard. 

 - [RabbitMQ](https://www.rabbitmq.com/) (with the [rabbitmq_amqp1_0](https://github.com/rabbitmq/rabbitmq-server/tree/master/deps/rabbitmq_amqp1_0) plugin enabled)
 - [Azure Service Bus](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview)

## Kubernetes 

See the [Helm chart](/chart/openfaas-amqp1.0-connector) for details

## Configuration

Configuration is achieved via environment variables:

| Variable | Description | Default |
|---|---|---|
| AMQP_ATTEMPT_RECONNECTION | Attempt reconnection if AMQP server disconnects | `true` |
| AMQP_DELIVERY_MAX_ATTEMPTS | Number of times to attempt delivery of a message in the event of failure | 3 |
| AMQP_HOST | Host of the AMQP server | `undefined` |
| AMQP_HOSTNAME | Hostname of the AMQP server (Azure Service Bus requires this to be same as `AMQP_HOST`) | `undefined` |
| AMQP_PASSWORD | Password for the AMQP server | `undefined` |
| AMQP_PORT | Port for the AMQP server | `5672` |
| AMQP_QUEUE_NAME | The name of the queue to subscribe to | `undefined` |
| AMQP_RECEIVER_AUTO_ACCEPT | Should messages be auto-accepted - this will prevent redelivery of messages on failure | `false` |
| AMQP_RECONNECT_LIMIT | Number of times to attempt reconnection | `1` |
| AMQP_RESPONSE_REPLY_QUEUE | The name of the queue to publish replies to | `${AMQP_QUEUE_NAME}_reply` |
| AMQP_RESPONSE_SEND_REPLY | Should replies be sent | `true` |
| AMQP_RECEIVER_FLOW_MANUAL | Should the receiver flow be manually controlled | `false` |
| AMQP_RECEIVER_FLOW_CONCURRENCY | The maximum number of items that can be processed at once. Will be ignored if `AMQP_RECEIVER_FLOW_MANUAL` if `false` | `500` |
| AMQP_RECEIVER_FLOW_POST_PROCESS_PAUSE | Pause after each item is processed. In milliseconds. | `0` |
| AMQP_TRANSPORT | Transport method - can be `tls`, `ssl` or `tcp` | `undefined` |
| AMQP_USERNAME | Username for the AMQP server | `undefined` |
| HEALTH_PORT | Port to run health checks on | `3000` |
| LOGGER_LEVEL | Reporting level for logs | `info` |
| OPENFAAS_ASYNC | The function is asynchronous - this will prevent replies being sent to a queue | `false` |
| OPENFAAS_CALLBACK_URL | The callback URL to send to an asynchronous function | `undefined` |
| OPENFAAS_FUNCTION | The OpenFaaS function to target | `undefined` |
| OPENFAAS_GATEWAY | The OpenFaaS gateway URL to target | `undefined` |
| OPENFAAS_PASSWORD | Password for the OpenFaaS gateway | `undefined` |
| OPENFAAS_USERNAME | Username for the OpenFaaS gateway | `undefined` |

## Worked Examples

These examples use the Docker image for simplicity. If you have NodeJS installed, you can build the package yourself:

```shell
npm run build
node .
```

### RabbitMQ

```shell
docker run -it --rm \
  -e OPENFAAS_FUNCTION=fn-name \
  -e OPENFAAS_GATEWAY=https://openfaas.example.domain \
  -e OPENFAAS_USERNAME=admin \
  -e OPENFAAS_PASSWORD=password \
  -e AMQP_HOST=rabbitmq-service \
  -e AMQP_USERNAME=admin \
  -e AMQP_PASSWORD=admin \
  -e AMQP_PORT=5672 \
  -e AMQP_QUEUE_NAME=example \
  registry.gitlab.com/mrsimonemms/openfaas-amqp1.0-connector
```

### Azure Service Bus

Azure Service Bus requires the `AMQP_HOST` and `AMQP_HOSTNAME` to be set to the same value.

```shell
docker run -it --rm \
  -e OPENFAAS_FUNCTION=fn-name \
  -e OPENFAAS_GATEWAY=https://openfaas.example.domain \
  -e OPENFAAS_USERNAME=admin \
  -e OPENFAAS_PASSWORD=password \
  -e AMQP_HOST=hostname.servicebus.windows.net \
  -e AMQP_HOSTNAME=hostname.servicebus.windows.net \
  -e AMQP_USERNAME=RootManageSharedAccessKey \
  -e AMQP_PASSWORD=my-password \
  -e AMQP_PORT=5671 \
  -e AMQP_TRANSPORT=tls \
  -e AMQP_QUEUE_NAME=example \
  registry.gitlab.com/mrsimonemms/openfaas-amqp1.0-connector
```

## Logging

Application logging is done using [Pino](https://getpino.io) and publishes to the console. To make the logs more readable,
pipe through [Pino Pretty](https://github.com/pinojs/pino-pretty) - `node . | pino-pretty -tlc`


CHANGELOG
# [1.1.0](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/compare/v1.0.1...v1.1.0) (2021-01-15)


### Features

* **throttling:** add receiverFlow options to control how quickly incoming messages are processed ([64afc17](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/64afc173bbdab09e6ecec06d4fbdb2eb1d997680))

## [1.0.1](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/compare/v1.0.0...v1.0.1) (2021-01-05)


### Bug Fixes

* **health:** change log level for healthcheck ([f5cc697](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/f5cc697d7ecfd74d1a3d8f1cf8eb0aa671652f9d))

# 1.0.0 (2020-12-27)


### Features

* configure class to interact with amqp server ([2aa5eff](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/2aa5eff4fefeb2a7fb63827b3e7fed503a10a9fa))
* configure health check for the amqp connection ([e433f0b](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/e433f0be1ce9923ee3b021f59d8f0f9580d55663))
* create class to manage invoking openfaas functions ([60aac07](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/60aac070794344494f515020176112bd4a91229a))
* create helm chart to manage deployment ([8760ad1](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/8760ad13c46234fe2c78aebd0a78bc91173b78f1))
* wire up openfaas, amqp and health checks as an application ([b84dce7](https://gitlab.com/MrSimonEmms/openfaas-amqp1.0-connector/commit/b84dce79ca92178932ccf58bb20c32c55537946e))


src/amqp.ts
/**
 * amqp
 */

/* Node modules */

/* Third-party modules */
import { Container, create_container, EventContext, Message } from 'rhea';
import { Logger } from 'pino';
import { ReceiverOptions } from 'rhea/typings/connection';

/* Files */
import { IAMQP, IAMQPConfig, IAMQPHealth, IOpenFaaS } from './interfaces';

export default class AMQP implements IAMQP {
  private connectionHealth: IAMQPHealth = {
    receiver: false,
    sender: false,
  };

  private container: Container;

  /* Used to send replies */
  private sender: EventContext | undefined;

  constructor(private config: IAMQPConfig, private logger: Logger, private openfaas: IOpenFaaS) {
    this.container = create_container();

    this.configureListeners();
    this.startListening();
  }

  addCredit(ctx) {
    if (this.config.receiverFlow.auto) {
      this.logger.debug('Receiver flow is automatic');
      return;
    }

    this.logger.info('Incrementing credit');

    const maxConcurrency = this.config.receiverFlow.concurrentItems;
    const availableCredit = ctx.receiver.credit;
    const creditToAdd = maxConcurrency - availableCredit;

    this.logger.info(
      {
        maxConcurrency,
        availableCredit,
        creditToAdd,
      },
      'Adding credit to message queue listener',
    );

    ctx.receiver.add_credit(creditToAdd);
  }

  connectionStatus(): IAMQPHealth {
    return this.connectionHealth;
  }

  protected configureListeners() {
    this.container
      .on('accepted', () => {
        this.logger.debug('Message accepted');
      })
      .on('connection_close', () => {
        this.logger.warn('Connection closed');
      })
      .on('connection_error', (ctx) => {
        this.logger.error({ ctx }, 'Connection error');
      })
      .on('connection_open', () => {
        this.logger.debug('Connection open');
      })
      .on('disconnected', ({ error }) => {
        this.logger.fatal({ err: error }, 'AMQP connection disconnected');
        this.connectionHealth.sender = false;
        this.connectionHealth.receiver = false;
      })
      .on('error', (err) => {
        this.logger.error(
          {
            err: {
              message: err.message,
            },
          },
          'AMQP general error',
        );
      })
      .on('message', (ctx: EventContext) => this.processNewMessage(ctx))
      .on('protocol_error', (ctx) => {
        this.logger.error({ ctx }, 'Protocol error');
      })
      .on('receiver_open', (ctx) => {
        this.logger.info('Receiver open');

        this.connectionHealth.receiver = true;

        this.addCredit(ctx);
      })
      .on('receiver_close', () => {
        this.logger.debug('Receiver closed');
        this.connectionHealth.receiver = false;
      })
      .on('receiver_error', ({ error }) => {
        this.logger.error({ err: error }, 'Receiver error');
        this.connectionHealth.receiver = false;
      })
      .on('sendable', (ctx) => {
        this.sender = ctx;
        this.connectionHealth.sender = true;
        this.logger.info('Reply queue now sendable');
      })
      .on('sender_close', () => {
        this.logger.debug('Sender closed');
        this.connectionHealth.sender = false;
      })
      .on('sender_error', ({ error }) => {
        this.logger.error({ err: error }, 'Sender error');
        this.connectionHealth.sender = false;
      });
  }

  protected async processNewMessage(ctx: EventContext): Promise<void> {
    const logger = this.logger.child({
      messageId: ctx.message.message_id,
      correlationId: ctx.message.correlation_id,
    });

    try {
      const message = AMQP.convertMessage(ctx.message);
      logger.info(
        {
          message,
          subject: ctx.message.subject,
          contentType: ctx.message.content_type,
          deliveryCount: ctx.message.delivery_count,
          created: ctx.message.creation_time,
        },
        'New message received',
      );

      const { contentType, data } = await this.openfaas.invoke(message);

      if (this.config.response.sendReply) {
        const replyMessage: Message = {
          content_type: contentType,
          body: data,
          correlation_id: ctx.message.correlation_id,
        };

        logger.info(
          {
            ...replyMessage,
            body: undefined,
          },
          'Sending reply to queue',
        );

        if (!this.sender) {
          logger.error('Queue is not sendable yet - requeuing');
          return;
        }

        this.sender.sender.send(replyMessage);
      }

      /* Acknowledge receipt - message will never be sent again */
      ctx.delivery.accept();

      logger.info('New message successfully processed');
    } catch (err) {
      logger.error({ message: err.message }, 'Message processing errored');

      if (ctx.message.delivery_count >= this.config.delivery.maxAttempts) {
        logger.error('Exceeded delivery attempts - rejecting');

        ctx.delivery.reject();
      } else {
        ctx.delivery.release({
          delivery_failed: true,
          message_annotations: {
            error: err.message,
          },
        });

        logger.info('Message released');
      }
    }

    const { postProcessPause } = this.config.receiverFlow;

    if (postProcessPause > 0) {
      this.logger.info({ postProcessPause }, 'Pausing after message processed');

      await new Promise((resolve) => setTimeout(resolve, postProcessPause));
    }

    this.addCredit(ctx);
  }

  private startListening(): void {
    const connection = this.container.connect(this.config.connection);
    const receiverConfig: ReceiverOptions = {
      autoaccept: this.config.receiver.autoaccept,
      source: this.config.receiver.source,
    };

    if (!this.config.receiverFlow.auto) {
      this.logger.debug(this.config.receiverFlow, 'Manually controlling receiver flow');

      receiverConfig.credit_window = 0;
    }

    connection.open_receiver(receiverConfig);

    if (this.config.response.sendReply) {
      this.logger.debug({ queue: this.config.response.replyQueue }, 'Reply will be sent');
      connection.open_sender(this.config.response.replyQueue);
    } else {
      this.logger.debug('No reply will be sent');
    }
  }

  static convertMessage(message: Message): any {
    /* I don't like using instanceof, but can't think of a better way of doing it */
    const messageContent =
      message.body?.content instanceof Buffer ? message.body.content.toString() : message.body;

    let formattedMessage: any;
    switch (message.content_type) {
      case 'application/json':
        formattedMessage = JSON.parse(messageContent);
        break;

      default:
        // No formatting - store as normal
        formattedMessage = messageContent;
        break;
    }

    console.log(formattedMessage);

    return formattedMessage;
  }
}


src/health.ts
/**
 * health
 */

/* Node modules */

/* Third-party modules */
import * as express from 'express';
import { Logger } from 'pino';

/* Files */
import { IAMQP, IHealthConfig } from './interfaces';

export default (config: IHealthConfig, logger: Logger, amqp: IAMQP) => {
  const app = express();

  app.get('/health', (req, res) => {
    const status = amqp.connectionStatus();

    const services = Object.keys(status).map((name) => ({
      name,
      isHealthy: status[name],
    }));
    const isHealthy = services.every((item) => item.isHealthy);

    logger.info({ isHealthy, status }, 'Check health of AMQP service');

    if (!isHealthy) {
      res.status(503);
      logger.warn({ services }, 'Service is unhealthy');
    }

    res.send({
      status: isHealthy ? 'ok' : 'error',
      services,
    });
  });

  app.listen(config.port, () => {
    logger.info('Health checks listening');
  });
};


src/interfaces.ts
/**
 * interfaces
 */

/* Node modules */

/* Third-party modules */
import { ConnectionOptions, ReceiverOptions } from 'rhea/typings/connection';

/* Files */

export interface IAMQPHealth {
  receiver: boolean;
  sender: boolean;
}

export interface IAMQP {
  connectionStatus(): IAMQPHealth;
}

export interface IAMQPConfig {
  connection: ConnectionOptions;
  delivery: {
    maxAttempts: number;
  };
  receiver: ReceiverOptions;
  response: {
    replyQueue: string;
    sendReply: boolean;
  };
}

export interface IHealthConfig {
  port: number;
}

export interface IOpenFaaSInvokeResult {
  contentType: string;
  data: any;
}

export interface IOpenFaaS {
  invoke(message: any): Promise<IOpenFaaSInvokeResult>;
}

export interface IOpenFaaSConfig {
  async: boolean;
  callbackUrl?: string;
  function: string;
  gateway: string;
  password?: string;
  username?: string;
}


src/config.ts
/**
 * config
 */

/* Node modules */

/* Third-party modules */
import { ConnectionOptions, ReceiverOptions } from 'rhea/typings/connection';

/* Files */

const queueName = process.env.AMQP_QUEUE_NAME ?? 'openfaas';

export default {
  logger: {
    level: process.env.LOGGER_LEVEL ?? 'info',
    redact: ['config.amqp.connection.password', 'config.openfaas.password'],
  },
  amqp: {
    connection: {
      host: process.env.AMQP_HOST,
      hostname: process.env.AMQP_HOSTNAME,
      reconnect_limit: Number(process.env.AMQP_RECONNECT_LIMIT ?? 1),
      password: process.env.AMQP_PASSWORD,
      port: Number(process.env.AMQP_PORT ?? 5672),
      reconnect: process.env.AMQP_ATTEMPT_RECONNECTION !== 'false',
      transport: process.env.AMQP_TRANSPORT,
      username: process.env.AMQP_USERNAME,
    } as ConnectionOptions,
    delivery: {
      maxAttempts: Number(process.env.AMQP_DELIVERY_MAX_ATTEMPTS ?? 3),
    },
    receiver: {
      autoaccept: process.env.AMQP_RECEIVER_AUTO_ACCEPT === 'true', // Default to "false"
      source: queueName,
    } as ReceiverOptions,
    response: {
      replyQueue: process.env.AMQP_RESPONSE_REPLY_QUEUE ?? `${queueName}_reply`,
      sendReply: process.env.AMQP_RESPONSE_SEND_REPLY !== 'false',
    },
  },
  health: {
    port: Number(process.env.HEALTH_PORT ?? 3000),
  },
  openfaas: {
    async: process.env.OPENFAAS_ASYNC === 'true',
    callbackUrl: process.env.OPENFAAS_CALLBACK_URL,
    function: process.env.OPENFAAS_FUNCTION ?? queueName, // Default to same as exchange
    gateway: process.env.OPENFAAS_GATEWAY,
    password: process.env.OPENFAAS_PASSWORD,
    username: process.env.OPENFAAS_USERNAME,
  },
};
